### DISCUSSION TITLE
Using <For /> to Generate children of Component

### DISCUSSION THREAD

**Discussion by jcuenod** [solidjs/solid#496] [Q&A]
I have a component that expects to receive a `[children]` array and interleaves them with other elements. In the parent, it sometimes makes sense to create the children with a `<For each={children} />`. The problem is that I'm getting a signal passed into the component now instead of an array of children. How do I handle that?

I tried

```js
const children = typeof props.children === "function"
    ? props.children()
    : props.children
```

But this doesn't seem to be working.

**Comment by ryansolid** [comment]
Use Solid's `children` helper: https://github.com/solidjs/solid/blob/main/documentation/api.md#children

The problem is every time you access `props.children` you are recreating the children. If you wrap with children it will flatten out all nested reactivity and give you a signal you can work with.

**Reply by giubatt** [reply]
It seems like the tutorial for children is broken due to the helper returning an acessor to a JSX.Element, and typescript complains that it doesn't have forEach

https://www.solidjs.com/tutorial/props_children?solved

**Reply by ryansolid** [reply]
I see yeah it may or may not be an array. That's fun. Open to suggestions.

**Reply by atk** [reply]
We could extend the helper to accept any JSX.Element and return an array form (undefined/null/false=[], singular Element=[Element], otherwise keep props.children); maybe even have an optional flag to resolve function components inside an extra handler.

**Reply by ryansolid** [reply]
I debated that but decided not wanting to take the overhead in the helper because I don't make similar concessions elsewhere. I meant are there TS suggestions? I know my thinking is backwards, I never want TS to affect what the runtime JS is. It's stubborn position but I still think in JavaScript, so when TS would have any impact on what I want to accomplish API or runtime-wise I'm inclined to blame its shortcomings rather than my unwillingness to play along.

**Reply by giubatt** [reply]
Maybe `children` should be able to receive an optional generic that will be the return type of the accessor, and it can be JSX.Element by default.

So if you know props.children is an array you can just use it like this
`const c = children<JSX.Element[]>(() => props.children);`
`c` would then be typed as `Accessor<JSX.Element[]>`

**Reply by ryansolid** [reply]
@giubatt yeah that sounds reasonable to me. If you type narrow so be it or just state it. Thanks.
